File: src/shared/ipc/dataset.ts
import { invoke } from '@tauri-apps/api/core';
import type { CreateVariableResult, LoadedDataset, VariableSpec } from '../types';

/** シート名の一覧を返す。空配列は「シート選択が不要な形式」の合図。 */
export async function getSheets(path: string): Promise<string[]> {
  return invoke<string[]>('get_sheets', { path });
}

export async function loadDataset(path: string, sheet?: string): Promise<LoadedDataset> {
  return invoke<LoadedDataset>('load_dataset', { path, sheet });
}

/** 派生列を追加し、更新後のデータセットを返す。キャッシュのキーは変わらない。 */
export async function createVariable(
  datasetKey: string,
  spec: VariableSpec
): Promise<CreateVariableResult> {
  return invoke<CreateVariableResult>('create_variable', { datasetKey, spec });
}

File: src-tauri/src/commands/dataset.rs
use std::path::PathBuf;

use tauri::State;

use crate::bootstrap::AppState;
use crate::models::{
    CreateVariableResult,
    LoadedDataset,
    VariableSpec,
};

#[tauri::command]
pub async fn get_sheets(path: String,
                        state: State<'_, AppState>)
                        -> Result<Vec<String>, String> {
    let dataset = state.dataset.clone();
    tauri::async_runtime::spawn_blocking(move || dataset.get_sheets(&PathBuf::from(path)))
        .await.map_err(|e| format!("シート一覧の取得失敗: {e}"))?
}

#[tauri::command]
pub async fn load_dataset(path: String,
                          sheet: Option<String>,
                          state: State<'_, AppState>)
                          -> Result<LoadedDataset, String> {
    let dataset = state.dataset.clone();
    tauri::async_runtime::spawn_blocking(move || dataset.load(&PathBuf::from(path), sheet))
        .await.map_err(|e| format!("データセットの読込失敗: {e}"))?
}

#[tauri::command]
pub async fn create_variable(dataset_key: String,
                             spec: VariableSpec,
                             state: State<'_, AppState>)
                             -> Result<CreateVariableResult, String> {
    let dataset = state.dataset.clone();
    tauri::async_runtime::spawn_blocking(move || dataset.create_variable(&dataset_key, &spec))
        .await.map_err(|e| format!("変数の作成失敗: {e}"))?
}
